### ISSUE TITLE
Leaking database connection issue

### ISSUE THREAD

**Issue by Galathius** [Envek/after_commit_everywhere#20]
Thank you so much for this gem. It is a really great thing to have in our code base.

After we added this gem to our project we started getting a lot of these kinds of errors from the Sidekiq:
```
  could not obtain a connection from the pool within 5.000 seconds (waited 5.000 seconds); all pooled connections were in use
```

I may assume that this is somehow relates to usage of `ActiveRecord::Base.connection` instead of `ActiveRecord::Base.connection_pool.with_connection { |connection| ... }`.

So far we have fixed that issue by wrapping all `after_commit { do_stuff }` instructions with:

```
ActiveRecord::Base.connection_pool.with_connection do |conn|
  after_commit(connection: conn) { ...do_stuff... }
end
```

Probably, it would be nice to have some guard [here](https://github.com/Envek/after_commit_everywhere/blob/master/lib/after_commit_everywhere.rb#L103) to check if we have any active connections, if not – we can run the block immediately (absolutely the same check as you have for an opened transaction).

Useful links:
- https://gitlab.com/gitlab-org/gitlab-foss/-/issues/27949

**Comment by Galathius** [comment]
Gem versions:
- activerecord (= 6.1.5.1)
- sidekiq (6.4.0)
- sidekiq-scheduler (3.1.0)

You're right, `ActiveRecord::Base.connection` reuses existing db-connection. But there may be a case when connection is not created yet for the current thread. (e.g. you have bunch of sidekiq jobs that are not using db at all, just do some requests to 3rd-pary service and schedules new jobs).
Probably, for more safety would be cool to use `ActiveRecord::Base.connection_pool.connected?`, just to not have any side-effects. 

More over, guys from Sidekiq added support of `after_commit_everywhere` [out of the box](https://github.com/mperham/sidekiq/pull/5291/files), this change may be relevant for them.

BTW, I'm still not 100% sure that `ActiveRecord::Base.connection` is a reason for leaking database connections in my project. I'm going to wait for few more days to be 100% sure (the issue is very flaky, but we faced this right after we released integration with `after_commit_everywhere`).

**Comment by Galathius** [comment]
And a bit more details about how we use this gem. (Almost the same as Sidekiq guys use it)

`lib/sidekiq/postpone_scheduling_until_transaction_commit.rb`:
```
module Sidekiq
  module PostponeSchedulingUntilTransactionCommit
    include AfterCommitEverywhere

    def raw_push(payloads)
      if payloads.first['postpone_scheduling_until_transaction_commit'] == false
        super
      else
        ActiveRecord::Base.connection_pool.with_connection do |conn|
          after_commit(connection: conn) { super }
        end
      end
    end
  end
end
```

`config/initializers/sidekiq.rb`
```
require 'sidekiq/postpone_scheduling_until_transaction_commit'
Sidekiq::Client.public_send(:prepend, Sidekiq::PostponeSchedulingUntilTransactionCommit)
```

**Comment by Envek** [comment]
Got it, thanks.

I think that it is better not to checkout new connection with `connection_pool.with_connection` (which maybe pretty expensive in case when there are no available existing connections and new connection need to be established and configured), but just execute block in-place.

I'm probably going to add `connection_pool.connected?` check proposed by you (but need to think on tests for this).

**Comment by Envek** [comment]
Ok, I think the reason why this gem worked well for years and you're literally the first person to complain is that usually all its invocations (whether it happens during serving HTTP request in Rails controller or performing job in Sidekiq worker process) are always done inside [Rails executor](https://guides.rubyonrails.org/threading_and_code_execution.html#executor) which checks in any connections back to the connection pool.
See https://guides.rubyonrails.org/threading_and_code_execution.html for details (thanks @palkan for the link)

It seems that sometimes your app enqueues Sidekiq jobs outside of executor context (not within controller action or another Sidekiq job) and in that case connection is being checked out and never checked back in.

So after_commit_everywhere should be fixed for such usages and non-Rails apps using ActiveRecord.

**Comment by Envek** [comment]
Here is the pull request, feel free to comment: https://github.com/Envek/after_commit_everywhere/pull/21

**Comment by Envek** [comment]
Fix has been released in version [1.2.1](https://github.com/Envek/after_commit_everywhere/releases/tag/v1.2.1).

Thank you very much for reporting!

**Comment by Galathius** [comment]
As soon as I give up with my workaround and updated gem version – the problem started to appear again :(
Seems `ActiveRecord::Base.connection_pool.connected?` is not a proper way to detect that we're currently in the "active" connection. `connected?` return true if the current process has at least one open connection (that may not be used right now).

For example, you can run this code from the rails console:
```
3.0.3 (main):0 > ActiveRecord::Base.connection_pool.connections.size
=> 0
3.0.3 (main):0 > 3.times.map { Thread.new { AfterCommitEverywhere.after_commit {puts 1} } }.map(&:join)
=> ....
3.0.3 (main):0 > ActiveRecord::Base.connection_pool.connections.size
=> 0
3.0.3 (main):0 > ActiveRecord::Base.connection
=> ...
3.0.3 (main):0 > ActiveRecord::Base.connection_pool.connections.size
=> 1
3.0.3 (main):0 > 3.times.map { Thread.new { AfterCommitEverywhere.after_commit {puts 1} } }.map(&:join)
=> ...
3.0.3 (main):0 > ActiveRecord::Base.connection_pool.connections.size
=> 4
```

Ideally, we need to find a way to check if there is any `open connection` without opening connection (what ActiveRecord::Base.connection` does. MAYBE we need to try `::Rails.application.executor.active?` method to identify should we close connection manually or not.

**Comment by Galathius** [comment]
Seems I found a way to reproduce it from Rails console. Just run each example in separate rails consoles:
```ruby
5.times.map do |i|
  Thread.new do
    sleep i
    ActiveRecord::Base.connection_pool.with_connection do |conn|
      AfterCommitEverywhere.after_commit(connection: conn) { 'do stuff' }
    end
  end
end.map(&:join)

ActiveRecord::Base.connection_pool.connections.count # => 1
```
Here everything works good, the single connection is reused for multiple thread.


```
5.times.map do |i|
  Thread.new do
    sleep i
    AfterCommitEverywhere.after_commit { 'do_stuff' }
  end
end.map(&:join)

ActiveRecord::Base.connection_pool.connections.count # => 0
```
The process's connections pool is not connected to the database, no new connections were created, OK.

```
ActiveRecord::Base.connection
5.times.map do |i|
  Thread.new do
    sleep i
    AfterCommitEverywhere.after_commit { 'do_stuff' }
  end
end.map(&:join)

ActiveRecord::Base.connection_pool.connections.count # => 6
```
Leaking connections.

**Comment by Envek** [comment]
Thanks for reproduction steps! I will take another look into it.

**Comment by Envek** [comment]
One more fix has been released in [1.2.2](https://github.com/Envek/after_commit_everywhere/releases/tag/v1.2.2) (using `active_connection?` instead of `connected?`)
